[PyTorch][mobile] Audit mobile interpreter for extra copies

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54076 [PyTorch] Fix string copy in WARN path for both interpreters
* **#54031 [PyTorch][mobile] Audit mobile interpreter for extra copies**

Similar to D27060762, caught some probably-unintended copies.

Differential Revision: [D27061818](https://our.internmc.facebook.com/intern/diff/D27061818/)

swolchok committed:
[PyTorch][mobile] Audit mobile interpreter for extra copies

Similar to D27060762, caught some probably-unintended copies.

Differential Revision: [D27061818](https://our.internmc.facebook.com/intern/diff/D27061818/)

[ghstack-poisoned]

iseeyuan reviewed (approved):
Looks great to me. Thanks @swolchok !

iseeyuan commented on torch/csrc/jit/mobile/interpreter.cpp:
```diff
@@ -191,8 +191,10 @@ bool InterpreterState::run(Stack& stack) {
       case NAMED_TUPLE_CONSTRUCT: {
-        auto type = code_->types_[inst.X]->expect<at::TupleType>();
-        namedTupleConstruct(stack, type, inst.N);
+        namedTupleConstruct(
```
Should we also do this in DICT_CONSTRUCT and LIST_CONSTRUCT?

swolchok replied:
No, `listConstruct` takes const reference to `ListType` (i.e., not a shared_ptr) and `dictConstruct` takes const reference to `DictTypePtr`. It does seem weird to have `namedTupleConstruct` and `dictConstruct` both taking shared_ptrs, but not agree on whether it's by value or by reference. After looking, probably `dictConstruct` should take const reference to`DictType` too.

swolchok committed:
revert sref change on "[PyTorch][mobile] Audit mobile interpreter for extra copies"


Similar to D27060762, caught some probably-unintended copies.

Differential Revision: [D27061818](https://our.internmc.facebook.com/intern/diff/D27061818/)

[ghstack-poisoned]

swolchok committed:
format on "[PyTorch][mobile] Audit mobile interpreter for extra copies"


Similar to D27060762, caught some probably-unintended copies.

Differential Revision: [D27061818](https://our.internmc.facebook.com/intern/diff/D27061818/)

[ghstack-poisoned]
